File: cli/commands/stats.py
import os
from typing import Dict, Any

from cli.utils.file_utils import count_solutions, get_all_cpp_files
from cli.utils.ui import (
    print_command_banner,
    print_error,
    print_info,
    print_list_item,
    print_section,
    print_warning,
    console,
)


def collect_stats(base_dir: str, data_structures: Dict[str, str]) -> Dict[str, Any]:
    files = get_all_cpp_files(base_dir)

    by_structure = {name: 0 for name in data_structures}
    unmatched_problems = 0
    total_solutions = 0

    structure_tokens = {
        name: f"{os.sep}{folder}{os.sep}"
        for name, folder in data_structures.items()
    }

    for file_path in files:
        normalized_path = os.path.normpath(file_path)
        normalized_lower = normalized_path.lower()

        matched = False
        for name, token in structure_tokens.items():
            if token.lower() in normalized_lower:
                by_structure[name] += 1
                matched = True
                break

        if not matched:
            unmatched_problems += 1

        try:
            with open(file_path, "r", encoding="utf-8") as file_obj:
                total_solutions += count_solutions(file_obj.read())
        except OSError:
            continue

    return {
        "total_problems": len(files),
        "total_solutions": total_solutions,
        "by_structure": by_structure,
        "unmatched_problems": unmatched_problems,
    }


def main(config: Dict[str, Any]):
    print_command_banner("Problem Statistics")

    base_dir = config.get("base_dir", "")
    data_structures = config.get("data_structures", {})

    if not base_dir:
        print_error("Missing 'base_dir' in config")
        return

    if not os.path.isdir(base_dir):
        print_warning(f"Base directory does not exist: {base_dir}")
        return

    stats = collect_stats(base_dir, data_structures)

    print_section("Overview")
    print_list_item("Base directory", base_dir)
    print_list_item("Total problems", str(stats["total_problems"]))
    print_list_item("Total solutions", str(stats["total_solutions"]))
    console.print()

    print_section("By Data Structure")
    if not data_structures:
        print_info("No data structures configured")
    else:
        for name in sorted(data_structures.keys()):
            print_list_item(name, str(stats["by_structure"].get(name, 0)))

    if stats["unmatched_problems"] > 0:
        console.print()
        print_info(
            f"Unmatched problems: {stats['unmatched_problems']}"
        )
    console.print()

File: go-cli/internal/commands/stats.go
package commands

import (
	"fmt"
	"os"
	"strings"

	"leetcli/internal/config"
	"leetcli/internal/template"
)

func Stats(args []string, cfg *config.Config, ui UI) {
	ui.WriteOutput(MsgPlain, "--- Problem Statistics ---\n")

	baseDir := cfg.BaseDir
	dataStructures := cfg.GetDataStructures()

	if baseDir == "" {
		ui.WriteOutput(MsgError, "Missing 'base_dir' in config")
		return
	}
	if _, err := os.Stat(baseDir); os.IsNotExist(err) {
		ui.WriteOutput(MsgError, "Base directory does not exist: %s", baseDir)
		return
	}

	languages := template.NormalizeLanguages(nil)
	for k, v := range cfg.Languages {
		languages[k] = template.LanguageInfo{Label: v.Label, Ext: v.Ext}
	}
	var exts []string
	for _, info := range languages {
		exts = append(exts, info.Ext)
	}

	files := template.GetAllSolutionFiles(baseDir, exts)
	totalProblems := len(files)
	totalSolutions := 0

	byStructure := make(map[string]int)
	for name := range dataStructures {
		byStructure[name] = 0
	}
	unmatched := 0

	for _, f := range files {
		matched := false
		for name, folder := range dataStructures {
			token := strings.ToLower("/" + folder + "/")
			if strings.Contains(strings.ToLower(f), token) {
				byStructure[name]++
				matched = true
				break
			}
		}
		if !matched {
			unmatched++
		}

		if content, err := os.ReadFile(f); err == nil {
			totalSolutions += template.CountSolutions(string(content))
		}
	}

	ui.WriteOutput(MsgPlain, "\nOverview:")
	ui.WriteOutput(MsgInfo, "Base directory: %s", baseDir)
	ui.WriteOutput(MsgInfo, "Total problems: %d", totalProblems)
	ui.WriteOutput(MsgInfo, "Total solutions: %d", totalSolutions)

	ui.WriteOutput(MsgPlain, "\nBy Data Structure:")
	if len(dataStructures) == 0 {
		ui.WriteOutput(MsgInfo, "No data structures configured")
	} else {
		for _, name := range sortedKeys(byStructure) {
			ui.WriteOutput(MsgPlain, "  %-20s %d", name, byStructure[name])
		}
	}

	if unmatched > 0 {
		ui.WriteOutput(MsgInfo, "Unmatched problems: %d", unmatched)
	}
}

func sortedKeys(m map[string]int) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	for i := 0; i < len(keys); i++ {
		for j := i + 1; j < len(keys); j++ {
			if keys[i] > keys[j] {
				keys[i], keys[j] = keys[j], keys[i]
			}
		}
	}
	return keys
}

var _ = fmt.Sprintf
